plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/stock-indicator/block.tsx

```diff
@@ -9,24 +10,79 @@ import {
 import { useStyleProps } from '@woocommerce/base-hooks';
 import { withProductDataContext } from '@woocommerce/shared-hocs';
 import type { HTMLAttributes } from 'react';
+import { ProductResponseItem } from '@woocommerce/types';
+import { getSetting } from '@woocommerce/settings';
+import { useSelect } from '@wordpress/data';
 
 /**
  * Internal dependencies
  */
 import './style.scss';
 import type { BlockAttributes } from './types';
+import { store as woocommerceTemplateStateStore } from '../../../../shared/store';
+import type { ProductTypeProps } from '../../../../utils/get-product-type-options';
 
 type Props = BlockAttributes & HTMLAttributes< HTMLDivElement >;
 
+/**
+ * Determines whether the stock indicator should be visible based on product type and availability.
+ *
+ * @param product                             The product.
+ * @param availabilityText                    The stock availability text.
+ * @param isDescendentOfSingleProductTemplate Whether this is descendent of single product template.
+ * @param selectedProductType                 The selected product type.
+ * @return True if stock indicator should be visible.
+ */
+const isStockIndicatorVisible = (
+	product: ProductResponseItem,
+	availabilityText: string,
+	isDescendentOfSingleProductTemplate: boolean | undefined,
+	selectedProductType: string | undefined
+) => {
+	const productTypesWithoutStockIndicator = getSetting< string[] >(
+		'productTypesWithoutStockIndicator',
+		[ 'external', 'grouped', 'variable' ]
+	);
+
+	const productType = selectedProductType || product?.type;
+
+	const isProductTypeAllowed =
+		! productTypesWithoutStockIndicator.includes( productType );
+
+	return (
+		isProductTypeAllowed &&
+		( availabilityText !== '' || isDescendentOfSingleProductTemplate )
```

Comment:
It's still not clear to me why we need `isDescendentOfSingleProductTemplate`. When we are in the Single Product template, we should be able to rely on `isProductTypeAllowed`; and when we are in the Single Product block, we should be able to rely on `availabilityText`, no? :thinking: